crypto/fipsmodule/rand/entropy/tree_drbg_jitter_entropy.c

```diff
@@ -0,0 +1,465 @@
+      jitter_generated_output = 1;
+      break;
+    }
+
+    // If Jitter entropy failed to produce entropy we need to reset it.
+    jent_entropy_collector_free(tree_jitter_drbg->jitter_ec);
+    tree_jitter_drbg->jitter_ec = NULL;
+    tree_jitter_drbg->jitter_ec = jent_entropy_collector_alloc(0, JENT_FORCE_FIPS);
+    if (tree_jitter_drbg->jitter_ec == NULL) {
+      abort();
+    }
+  }
+
+  if (jitter_generated_output != 1) {
+    abort();
+  }
+}
+
+// tree_jitter_drbg_maybe_get_pred_resistance generates RAND_PRED_RESISTANCE_LEN
+// bytes for prediction resistance and returns them in |pred_resistance|.
+// However, it only generate bytes if |tree_jitter_drbg| meets the conditions:
+// 1) is not the global seed DRBG 2) is not protected from UBEs. If bytes are
+// generated, |pre_resistance_len| is set to RAND_PRED_RESISTANCE_LEN and is
+// otherwise not mutated.
+static void tree_jitter_drbg_maybe_get_pred_resistance(
+  struct tree_jitter_drbg_t *tree_jitter_drbg,
+  uint8_t pred_resistance[RAND_PRED_RESISTANCE_LEN],
+  size_t *pred_resistance_len) {
+
+  if (tree_jitter_drbg->is_global == 0 &&
+      tree_jitter_drbg->ube_protection != 1) {
+    CRYPTO_sysrand_for_seed(pred_resistance, RAND_PRED_RESISTANCE_LEN);
+    *pred_resistance_len = RAND_PRED_RESISTANCE_LEN;
+  }
+}
+
+// tree_jitter_check_drbg_must_reseed computes whether |state| must be
+// randomized to ensure uniqueness.
+//
+// Return 1 if |state| must be randomized. 0 otherwise.
+static int tree_jitter_check_drbg_must_reseed(
+  struct tree_jitter_drbg_t *tree_jitter_drbg) {
+
+  uint64_t current_generation_number = 0;
+  if (CRYPTO_get_ube_generation_number(&current_generation_number) != 1 &&
+      current_generation_number != tree_jitter_drbg->generation_number) {
+    tree_jitter_drbg->generation_number = current_generation_number;
+    return 1;
+  }
+
+  if (tree_jitter_drbg->generate_calls_since_seed + 1 > tree_jitter_drbg->max_generate_calls) {
+    return 1;
+  }
+
+  return 0;
+}
+
+// tree_jitter_drbg_derive_seed generates a CTR_DRBG_ENTROPY_LEN byte seed from
+// the DRBG configured in |tree_jitter_drbg|. The generated bytes are returned
+// in |seed_out|.
+//
+// |tree_jitter_drbg_derive_seed| automatically handles reseeding the
+// associated DRBG if required. In addition, if UBE detection is not supported
+// prediction resistance is used to ensure bytes are generated safely.
+static void tree_jitter_drbg_derive_seed(
+  struct tree_jitter_drbg_t *tree_jitter_drbg,
+  uint8_t seed_out[CTR_DRBG_ENTROPY_LEN]) {
+
+  if (tree_jitter_drbg == NULL) {
+    abort();
+  }
+
+  if (tree_jitter_check_drbg_must_reseed(tree_jitter_drbg) == 1) {
+    uint8_t seed_drbg[CTR_DRBG_ENTROPY_LEN];
+    if (tree_jitter_drbg->is_global == 1) {
+      tree_jitter_get_root_seed(tree_jitter_drbg, seed_drbg);
+    } else {
+      CRYPTO_STATIC_MUTEX_lock_write(global_seed_drbg_lock_bss_get());
+      tree_jitter_drbg_derive_seed(*global_seed_drbg_bss_get(), seed_drbg);
```

Comment:
Yeah depth is hard-coded. It was sufficient for what we need :)

Reply:
Is there a better/more obvious way to enforce this? Such `tree_jitter_drbg == global_seed_drbg_bss_get()` use the root, otherwise call `tree_jitter_drbg_derive_seed(*global_seed_drbg_bss_get(), seed_drbg)`. Nothing enforces only a single thing has `is_global` set to one, but only a single thing can be returned from `global_seed_drbg_bss_get()`.

Reply:
Correct, nothing "enforces" that. Although it's only set in a function that can only be called once `tree_jitter_initialize_once()`.

Using `tree_jitter_drbg == *global_seed_drbg_bss_get()` is a good idea here. But I use the `is_global` field a few places and think it's better to stay consistent with that.